### PR TITLE
feat: [PL-41387]: OIDC Token Library Access Token Exchange Implementation

### PR DESCRIPTION
**Pull request by Raghav-Murali**
## Describe your changes

This PR implements the Access Token exchange part for the OIDC Token Library.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54210)

### CODE REVIEW

**Comment by piyush-harness** [comment]
*[`930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenConfigStructure.java` line 13 at r1](https://reviewable.io/reviews/harness/harness-core/54210#-NgvNBao3l3C-J5ibcfw:-NgvNBap-qc98qf9zUL6:bo4xkdy) ([raw file](https://github.com/harness/harness-core/blob/eb880c8708f68df9794d9c5e518c4b3a7a5eb2e0/930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenConfigStructure.java#L13)):*
> ```Java
> 
> public class OidcAccessTokenConfigStructure {
>   @JsonProperty(io.harness.oidc.accesstoken.OidcAccessTokenConstants.ACCESS_TOKEN_EXCHANGE_ENDPOINT)
> ```

nit: Static imports

**Comment by piyush-harness** [comment]
*[`930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenExchangeApi.java` line 18 at r1](https://reviewable.io/reviews/harness/harness-core/54210#-NgvNtKJ7AtO-C9HfAtH:-NgvNtKJ7AtO-C9HfAtI:b-755yl4) ([raw file](https://github.com/harness/harness-core/blob/eb880c8708f68df9794d9c5e518c4b3a7a5eb2e0/930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenExchangeApi.java#L18)):*
> ```Java
>   @POST("v1/token")
>   @Headers("Content-Type: application/json")
>   Call<io.harness.oidc.accesstoken.OidcAccessTokenResponse> exchangeToken(
> ```

Why fully qualified imports?

**Review by Raghav-Murali — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54210)* status: 8 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenConfigStructure.java` line 13 at r1](https://reviewable.io/reviews/harness/harness-core/54210#-NgvNBao3l3C-J5ibcfw:-NgwOLGa1HmbxC9SgryU:bb6tyiv) ([raw file](https://github.com/harness/harness-core/blob/eb880c8708f68df9794d9c5e518c4b3a7a5eb2e0/930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenConfigStructure.java#L13)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

nit: Static imports
</blockquote></details>

DONE
___
*[`930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenExchangeApi.java` line 18 at r1](https://reviewable.io/reviews/harness/harness-core/54210#-NgvNtKJ7AtO-C9HfAtH:-NgwOMuN8EsACEX20IIl:b-1g72gl) ([raw file](https://github.com/harness/harness-core/blob/eb880c8708f68df9794d9c5e518c4b3a7a5eb2e0/930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenExchangeApi.java#L18)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Why fully qualified imports?
</blockquote></details>

It was added by the IDE and I missed it. Have updated it now.

**Comment by Raghav-Murali** [comment]
trigger uts

File: 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenUtility.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    // Make the POST request and handle the response
+    Call<OidcWorkloadAccessTokenResponse> call = oidcAccessTokenExchangeApi.exchangeToken(oidcAccessTokenRequest);
+
+    try {
+      Response<OidcWorkloadAccessTokenResponse> response = call.execute();
+      if (response.isSuccessful()) {
+        OidcWorkloadAccessTokenResponse oidcAccessTokenResponse = response.body();
+        return oidcAccessTokenResponse;
+      } else {
+        log.error("Error encountered while exchanging OIDC Access Token");
```
Comment (gn-harness) [inline 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenUtility.java:57]: Add some additional metadata to these error logs

**Review by gn-harness — APPROVED** [review]

File: 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenConfigStructure.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import static io.harness.oidc.accesstoken.OidcAccessTokenConstants.ACCESS_TOKEN_IAM_SA_CREDENTIALS_ENDPOINT;
+import static io.harness.oidc.accesstoken.OidcAccessTokenConstants.ACCESS_TOKEN_STS_ENDPOINT;
+import static io.harness.oidc.accesstoken.OidcAccessTokenConstants.END_POINT;
+import static io.harness.oidc.accesstoken.OidcAccessTokenConstants.WORKLOAD_ACCESS_TOKEN_CONFIG;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class OidcAccessTokenConfigStructure {
+  @JsonProperty(ACCESS_TOKEN_STS_ENDPOINT) private OidcAccessTokenStsEndpoint oidcAccessTokenStsEndpoint;
+  @JsonProperty(ACCESS_TOKEN_IAM_SA_CREDENTIALS_ENDPOINT)
```
[the remark this answers is not in the document]
Comment (Raghav-Murali) [reply 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenConfigStructure.java:19]: Done

File: 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenUtility.java (hunk 1 above)
Comment (Raghav-Murali) [reply 930-oidc-core/src/main/java/io/harness/oidc/accesstoken/OidcAccessTokenUtility.java:57]: Done

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger uts

**Comment by Raghav-Murali** [comment]
trigger ut5, ut9

**Comment by Raghav-Murali** [comment]
trigger ut9

**Review by gn-harness — APPROVED** [review]

**Pull request merged**